#1&rpar; Vector databases and chroma


## 1.1 What kind of datastore are vector databases?

Vector databases are a highly specialized type of NoSQL databases that are designed to efficiently store and search high-dimenstional vectors. Instead of storing data in rows and columns, they are stored as vector embeddings, usually transformed from the original data.


##1.2 Why / When do we use them?
 Since vector databases work by turning unstructured data into embedding vectors, tasks such as searches and making recommendations and analysis can be implemented via similarity-based vector search.

 Take the example of searching, Traditional DBs (Mongo,SQL) searched for the exact word of "cookie". If a user searched for "baked sweet treat", it finds nothing. But a Vector DB understands the concept of "baked sweet treat" through semantic similarity search algorithm, that mathematically places "cookie" very close to "baked sweet treat" in the embedding space.

 When to use: Vector DBs have the follwing applications like -

*   LLM's - The rise of LLMs for tasks like information retrieval, alongside the increased popularity of e-commerce and recommendation platforms, requires vector database management systems that can deliver query optimization capabilities for unstructured data.
*   Image recognition - Vector databases can play a key role in image recognition by storing high-dimensional embeddings of images generated by ML models. As vector databases are optimized for similarity search tasks, this makes them ideal for applications such as object detection, facial recognition and image search.

*   Recommendation systems and personalization -
Once a vector database is trained using an embedding model, it can be utilized to generate personalized recommendations. When a user interacts with the system, their behavior and preferences are used to generate the user’s embedding





## 1.3 What is Chroma?
Chroma is an open source, AI-native vector database, designed for efficiently storing, searching and managing vector embeddings.If focuses heavily on the Developer experience, making it incredibly easy to plug knowledge, facts and skills to LLM applications without the need to manage complex database infrastructure.

##1.4 Chroma vs Milvus vs Weaviate vs Pinecone

###1.4.1 **Chroma**


*   **Easy API**: Chroma has a minimalistic API. It stores vectors with optional metadata and lets you do simple similarity queries. Setup is trivial (just pip install and point to a directory or let it manage storage). There’s no schema or explicit indexing step — just insert and query.

*   **Integration**: Chroma’s main selling point is glue with LangChain, LlamaIndex, and AI frameworks. Out of the box it has clients for Python and JavaScript, and advertises connectors for LangChain, LlamaIndex, OpenAI, Hugging Face, etc. If you’re building an LLM app in Python, Chroma just plugs right in.
*   **Managed vs OSS**: Chroma itself is Apache-licensed and free. It also overs Chroma cloud - a serverless hosted version with usage-based pricing, which cost a fraction of a cent. (charged per GB/written/stored/queried)

*  **Performance**: For small to medium datasets (millions of vectors), Chroma is fast and memory-efficient.  It’s best for “tens of millions” of vectors on a single machine.

*   **Tradeoffs**: It trades raw power for simplicity. It lacks advanced indexing options. It offers only HNSW, no GPU acceleration, so for GPU search, millions of writes/seconds or shrads, a heavier DB might be required.
*   **Approach** : It uses SQLite for metadata and hnswlib for its core indexing. It focuses on local-first development.




###1.4.2 **Milvus**

*   **Architecture**: Milvus offers both standalone and distributed modes. It can shard data, replicate for high availability, and use CPU/GPU acceleration. It supports multiple index types: IVF, HNSW, ANNOY, etc., so you can tune for speed or memory. Hybrid search (vector + scalar filtering) is fully supported.

*   **Performance**: In raw throughput, Milvus can be blazingly fast, especially if you add GPUs. Some benchmarks show Milvus achieving the best raw search speeds when properly configured. It can handle very large collections (20M+ vectors per “compute unit” in Zilliz Cloud’s terms).

*   **Integration**: It has a Python SDK (pymilvus), Go/Java clients, etc. Milvus integrates with Kubeflow, and LangChain/LlamaIndex support it natively. Being open-source, it’s often embedded in custom AI stacks.

*  **Trade-offs**: Milvus can be complex to operate and has a steep learning curve unlike Chroma. Its many moving parts (IndexServers, Proxy, DataCoord, etc.) make self-hosting non-trivial. It also tends to be resource-intensive at scale. If you don’t need billions of vectors or fancy indexes, this might be overkill. But for massive-scale AI search (enterprise RAG across huge corpora), it’s a solid choice.


*  **Approach** : Cloud - native architecture that completly decouples storage from computation, allowing them to scale independently.







###1.4.3 Weaviate

*  **Special features:** Weaviate has a GraphQL interface and treats objects as nodes in a knowledge graph. You define classes and properties, and Weaviate can index text, images, etc., into vectors on the fly using modules. Out-of-the-box integrations exist for Cohere, OpenAI, Hugging Face, etc., letting the DB itself generate embeddings. It also supports hybrid search (vector + keyword filters).

*   **Performance & scale:** Weaviate supports clustering (via Kubernetes) and HNSW indexing. It’s reasonably fast, but users report that complex GraphQL queries or very large schemas can add overhead. Weaviate Cloud (“WCS”) offers auto-scaling.

*   **Pricing/Deployment:** Weaviate Core is open-source (BSD). You can self-host with Docker/K8s. The fully-managed Weaviate Cloud is serverless pay-as-you-go: starts at 25 dollars/mo and ~0.095 dollars per million stored vector dimensions. (For 1M vectors at 1536 dims, that’d be ~145 dollars/mo plus base.) There’s also a free 14-day sandbox. Enterprise support and single-tenant clouds are available at higher tiers.

*   **Integration:** Weaviate provides REST and GraphQL APIs. It has official clients (Python, JS, Go). LangChain/LlamaIndex support Weaviate. The ability to auto-embed and connect to OpenAI/Cohere makes it popular in RAG pipelines.

*   **Tradeoffs** -  The schema/GraphQL approach is powerful but means more upfront design. There is a learning curve if you’re used to simple key/value stores. And because it’s Java under the hood, it often needs more RAM. If you want simplicity, Weaviate may be overkill.


*   **Approach** - Object-oriented structure accessed heavily via GraphQL




###1.4.4 Pinecone


*  **Scalability**: Full managed SaaS with zero infrastructure to manage. Can handle billions of vectors with low latency. It separates storage from coumpute, uses approximate search indexes, and offer a serverless experience. It offers auto - scaling, means you pay per query per storage. It is build for horizontal scale.

*   **Performance:** Benchmarks and reports consistently praise Pinecone’s raw query speed and reliability at scale. Typical latencies are low single-digit tens of ms for 10⁵–10⁶ vectors. Pinecone also supports hybrid queries (vector + metadata filter).
*   **Pricing (Managed):** Pinecone is proprietary. It offers a free Starter plan for small projects and prototypes. For production, the Standard plan has a 50 dollars/month minimum usage, then pay-as-you-go. (That breaks down to about 0.096 dollars per hour for an “s1.x1 pod” equivalent.) Enterprise plan requires 500 dollars /month minimum. Storage and query usage are metered. In short: higher cost than self-hosting, but predictable and fully managed.

* **Deployment:** Pinecone runs on AWS/GCP/Azure (you choose region). No on-premises or self-hosted option (except Pinecone Dedicated, which lets you deploy a private cloud region with their tech).

*   **Integration**: Pinecone provides REST and gRPC APIs, Python/Java/Go SDKs. It plugs into frameworks like LangChain and LlamaIndex (both list “Pinecone” as a first-class vectorstore). Pinecone also offers vector embedding and chat “Assistant” services (optional).

*  **Trade-offs:** Pinecone's biggest appeal is no maintainance. The downside is cost - you pay for simplicity and also gets expensive at scale. Also,customization is limited, closed-source and proprietary. There is no local, offline version for developers.
*   **Approach**: Proprietary serverless architecture optimized for real-time, low latency enterprise workloads.



## 1.5 Different modes in which Chroma can run



1.   **Ephemeral** (In -memory) mode : Rund entirely in-memory with no persistence. Use `chromadb.client()`
2.   **Persistent Client**: Saves and loads data from disk locally.
` chromadb.PersistentClient(path = "/path/to/save/to")`
3. **Client - Server Mode** : Chroma server runs as a separate process, clients connect via HTTP. chroma run --path/db-path and connect with HttpClient.
4. **Chroma cloud** : Serverless cloud-hosted database.
Use `chromadb.CloudClient()`





# 2&rpar; SETUP

Installing chromadb using pip...

In [ ]:
! pip install -q -U chromadb

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.0/52.0 kB 2.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.6/21.6 MB 86.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 278.2/278.2 kB 16.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 81.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.2/17.2 MB 107.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 72.1/72.1 kB 4.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 142.0/142.0 kB 9.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 68.7/68.7 kB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 231.6/231.6 kB 16.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 71.6/71.6 kB 4.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.6/60.6 kB 4.2 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the s

Using python, creating an ephemeral, in-memory, local client.

In [ ]:
import chromadb
client = chromadb.Client()

#3&rpar; Data creation

Creating a document database to hold vector embeddings for the 'Le cookie Quantique' database.

In [ ]:
collection = client.get_or_create_collection(name="answers")
doc1 = " Le Cookie Quantique is a premium galatic bakery specializing in cosmic - themed treats."
doc2 = " Our best selling item is the Black - Hole Brownie, which contains no nuts"
doc3 = " The Nebula Nutella Sablé is a fan favourite and contains hazelnuts and chocolate."
doc4 = " We are open Monday through Sunday from 08:00 AM to 06:00 PM Earth Time."
doc5 = " All of our Macaron de Mars are naturally gluten-free and made with almond flour."
doc6 = " We offer a fully vegan Supernova Soufflé for our plant-based space travellers."
doc7 = " Our flagship store is located on Earth, but we deliver anywhere in the solar system."
doc8 = " Our Croissant of the Cosmos is baked fresh daily at 4:00 AM to ensure maximum flakiness."

documents = [doc1,doc2,doc3,doc4,doc5,doc6,doc7,doc8]

query1 = "Do you have any vegan options?"
query2 = "When does the bakery open?"

document_ids = ["doc1","doc2","doc3","doc4","doc5","doc6","doc7","doc8"]

collection.add(
    ids = document_ids,
    documents=documents
)
print(f"Successfully added {collection.count()} documents to the '{collection.name}' collection!")


/root/.cache/chroma/onnx_models/all-MiniLM-L6-v2/onnx.tar.gz: 100%|██████████| 79.3M/79.3M [00:02<00:00, 32.0MiB/s]


Successfully added 8 documents to the 'answers' collection!


#4&rpar; Embedding Creation

##4.1 What does it mean to create vector embeddings and why do we do it?

Vector embeddings are numerical representations of data that capture the meaning or features of objects (like words, images, or concepts) as points in a multi-dimensional space, allowing machines to process and compare them efficiently. In essence, they are a way to translate complex information, such as words, sentences, images, or any other type of data, into lists of numbers that capture the underlying meaning and relationships within that data. These mathematical representations of data mimic human understanding, enabling a wide range of powerful artificial intelligence applications.

Training models to output vector representations of data points that correspond meaningfully to their real-world features enable us to make useful assumptions about how vector embeddings relate to one another. Intuitively, the more similar two real-world data points, the more similar their respective vector embeddings should be. Features or qualities shared by two data points should be reflected in both of their vector embeddings. Dissimilar data points should have dissimilar vector embeddings.

Armed with such logical assumptions, vector embeddings can be used as inputs to models that perform useful real-world tasks through mathematical operations that compare, transform, combine, sort or otherwise manipulate those numerical representations.

Vector embeddings thus underpin nearly all modern machine learning, powering models used in the fields of NLP and computer vision, and serving as the fundamental building blocks of generative AI.



##4.2 What is an embedding space?

An embedding space is essentially a mathematical, multi-dimensional grid where these vectors are plotted. In this space, concepts that are semantically similar (like "dog" and "puppy", or "cookie" and "brownie") are plotted physically close together, while concepts unrelated are plotted apart.

Expressing data points as vectors also enables the interoperability of different types of data, acting as a lingua franca of sorts between different data formats by representing them in the same embedding space. For example, smartphone voice assistants “translate” the user’s audio inputs into vector embeddings, and in turn use those vector embeddings for natural language processing (NLP) of that input.

## 4.3 So what happens when we use collections.add ?


*   By default, Chroma uses a **Sentence Transformers** `all-MiniLM-L6-v2` model to create embeddings.
*   This is a sentence transformer model that creates sentences and documents embeddings, numeric representation of your text that captures semantic meaning. When we call. add() function with documents, Chroma runs this model locally on our machine to convert our strings into dense vectors, which are then stored and indexed so that we can efficiently search for similar content.



Below is the code to count the size and dimensionality of local embedding space.

In [ ]:
num_documents = collection.count()
print(f"Number of documents in the 'answers' collection : {num_documents}")
test_vector = collection._embedding_function(["test"])
dimensionality = len(test_vector[0])
print(f" Dimensionality of the embedding space: {dimensionality}")

Number of documents in the 'answers' collection : 8
 Dimensionality of the embedding space: 384


###4.3.1 What does the number '384' represent?

The 384 represents the dimensionality of the mebedding vector, meaning each piece of text is converted into a vector of 384 floating point numbers. This is determined by the `all-MiniLM-L6-v2` 's model architecture. Each dimension captures some aspect of the text's semantic meaning, and together the 384 values form a point in high-dimensional space where semantically similar texts are positioned closer together.


###4.3.2 Does it match expectations?
It's a nested list: one inner list of 384 float values (since `all-MiniLM-L6-v2 `produces 384-dimensional embeddings) for each input string. The default algorithm Chroma uses is explicitly designed and documented by its creators to output 384-dimensional dense vectors.

# 5&rpar; Index Creation

##5.1 What is vector indexing?

With millions (and maybe billions) of vectors stored in a vector database, finding similarities through traditional brute-force searching can be slow ,*O(N)* . To speed searches, a common solution is to create a vector index of the database. Vector indexes speed vector search by using mathematical approximations rather than exact matches to find the best semantic results in the dataset.

The database maps out "neighbourhoods" of similar vectors so it can quickly zoom in on the relevant area of the embedding space and ignore the rest.

The most common approaches use approximate nearest neighbor searches, and there are a number of common algorithms that can be used, including Hierarchical Navigable Small World (HNSW), Product Quantization (PQ), and Inverted File Index (IVF).



##5.2 What are the main tradeoffs for precomputing vector indexes?

The choice between exact brute-force search and approximate indexing in vector databases involves balancing speed, memory usage, and accuracy. Brute-force methods guarantee perfect accuracy by comparing every query against all vectors, but they scale poorly with dataset size. Approximate indexes prioritize speed and memory efficiency by trading some accuracy, using techniques like graph-based or tree-based structures to narrow the search space.

The main tradeoff is Speed vs. Accuracy (Recall) vs. Memory.


*   Speed : Precomputing an index makes querying blisteringly fast from O(N) to O(logN), while it does not gurantee finding the absolute nearest neighbour. So there is a small tradeoff for accuracy for this speed.
*    Memory: Pre-computed structures, such as HNSW's graph edges or IVF's inverted lists, allow the system to quickly navigate to relevant data points without reading the entire dataset from disk but the index itself requires substantial RAM, sometimes 30–50% more than the raw vector data, particularly for graph-based algorithms that need to store links between nodes.


*   Once built, the index allows high-throughout querying, which is crucial for recommendation engines and RAG applications. But constructing this index is expensive. If the data changes rapidly, constant re-indexing can cripple performance.



##5.3 Write Ahead Log (WAL)

A Write-Ahead-Log (WAL) is an append-only file where a database records a new transaction before it actually applies that change to the main database or index. It is crucial for durability and crash recovery. If the server loses power mid-transaction, the database can simply read the WAL upon reboot to restore the lost data.

### 5.3.1 What is Chroma doing when it writes to the WAL?

As building a vector index is mathematically heavy, Chroma writes raw documents and embeddings to WAL to safely save them instantly. It gives the green light to the application, saying that the data is saved, while it takes time updating the complex vector in the background. Subsequently, immediately after the WAL update, the data in also written to the index, which enables Chroma to serve as a real-time search engine, where the data is readily available for querying after being writtten to the WAL.


###5.3.2 What algorithm did Chroma use to build this index?
Chroma automatically uses an algorithm called HNSW (Hierarchical Navigable Small World), specifically via the `hnswlib` library


###5.3.3 What does HNSW do?

HNSW is a graph-based index that efficiently finds approximate nearest neighbors in high-dimensional vector spaces. It builds a multi-layered graph where sparse upper layers act as "highways" for fast navigation, and during search it starts at the top and refines results as it moves down through denser layers. Chroma uses it in Single Node collections to quickly find the most similar embeddings to a query.

#6&rpar; Similarity Search

Vector similarity search is a technique that finds the most relavant data not by looking for exact keyword, but my measuring the mathematical distance between vectors. Because semantically similar concepts are clustered together in the multi-dimensional embedding space, the database can calculate metrics (like Cosine Similarity or Euclidean Distance) to find the document vectors that are physically closest to the query vector.

## 6.1 What does it mean to embed a query

Embedding a query means taking the user's raw text question and passing it through the exact same machine learning model used to embed the documents (in our case, `all-MiniLM-L6-v2)`. This translates the question into a 384-dimensional vector, placing it into the exact same mathematical space as the knowledge base so the database can measure the distance between the question and the potential answers.

### 6.1.1 Below we are running a similarity search by using .query for each of the 2 query strings and let us retrieve the top 2 documents.

In [ ]:
print("--- Query 1: Vegan Options ---")
results_1 = collection.query(
    query_texts = [query1],
    n_results = 2
)

for i,doc in enumerate(results_1['documents'][0]):
  print(f"Result {i+1}: {doc}")

print("--- Query 1: Opening time ---")
results_2 = collection.query(
    query_texts = [query2],
    n_results = 2
)

for i,doc in enumerate(results_2['documents'][0]):
  print(f"Result {i+1}: {doc}")

--- Query 1: Vegan Options ---
Result 1:  We offer a fully vegan Supernova Soufflé for our plant-based space travellers.
Result 2:  The Nebula Nutella Sablé is a fan favourite and contains hazelnuts and chocolate.
--- Query 1: Opening time ---
Result 1:  We are open Monday through Sunday from 08:00 AM to 06:00 PM Earth Time.
Result 2:  Le Cookie Quantique is a premium galatic bakery specializing in cosmic - themed treats.


###6.1.2 Interpretation of the result

The results represent the top n pieces of context from our database that are most semantically related to the user's question. For Query 1, the AI successfully found the exact vegan match for Result 1, but for Result 2, it grabbed the Nebula Nutella Sablé. Even though the Sablé isn't vegan, the AI must have mathematically recognized that both sentences belong to the same conceptual cluster ("baked goods/ingredients"), so it returned it as the next closest neighbor in the embedding space. And for query 2, it returned the accurate at first and then gave a result about the bakery itself, it must have been embedded such that all the general information regarding the bakery is embedded closely.

### 6.1.3 How to use it for chatbot?

For a chatbot, I would use a technique called RAG (Retrieval-Augmented Generation). I take these retrieved documents and inject them into a prompt for a Large Language Model (like Gemini or ChatGPT), effectively saying: "Answer the user's question, but strictly base my answer on these specific facts about Le Cookie Quantique." This prevents the AI from hallucinating and grounds its answers in my company's actual data.

#7&rpar; Scale

##7.1 What are the options to scale the database?

To handle massive expansion of datatypes (images, videos, audio) and volume, Le Cookie Quantique has a few primary options:


*   Migrate to a Distributed Architecture : We must move away from an ephemeral, in-memory Chroma setup. We would need to either deploy Chroma in a distributed Client/Server model across multiple cloud instances or migrate to an enterprise-grade vector database built specifically for horizontal scaling, such as Milvus or Pinecone.

*   Implement Multi - Modal Embeddings: To handle images, GIFs and song we cannot use our current text-only `all-MiniLM-l6-v2` model. We would need to upgrade to a multimodel embedding (like OpenAI's CLIP or Google's Vertex AI multi-modal embeddings). These models can project text, images, and audio into the exact same embedding space, allowing a user to upload a picture of a cookie and search our database for the matching recipe text.

* Database Sharding and Replication: As the company acquires competitors, the vector index will grow too large for a single machine's RAM. We would need to shard the database (split the index across multiple servers) and use read-replicas to handle the high volume of incoming customer queries.





##7.2 What considerations/tradeoffs do we need to weigh?



*   Infrastructure Management vs. Cost (The SaaS Tradeoff): Using a fully managed service like Pinecone means zero DevOps work, but it becomes extremely expensive at massive scale. Hosting our own distributed Milvus cluster on Kubernetes is much cheaper for compute costs, but requires hiring specialized engineers to maintain it.
*   Speed vs Accuracy (Index tuning): As the dataset grows into millions, searching becomes slower. We will have to tune our HNSW index parameters. We can make the search incredibly fast by limiting the search scope, but the tradeoff is that we might occasionally miss the absolute best matching product (lower recall accuracy).


*   Compute Heaviness of Multi-Modal models: Embedding an image of a song requires significantly more GPU compute power and time than embedding a simple sentence. We have to weigh the latency of how long a customer is willing to wait for the chatbot to reply against the cost of running high-end GPUs to process their multi-modal inputs instantly.



#8&rpar; Milvus VS Chroma

Let's use Milvus for creating our chatbot and see how the both compares.
We will start off by installing Milvus using pip.

## 8.1 Install pymilvus

In [ ]:
!pip install -U pymilvus

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 312.8/312.8 kB 5.9 MB/s eta 0:00:00


Next, we are setting up our connection to Milvus database through the cluster end point url and api token.

## 8.2 Establising the connection to Milvus DB

In [ ]:
from pymilvus import MilvusClient
from google.colab import userdata

CLUSTER_ENDPOINT = userdata.get('endpoint_url')
API_TOKEN = userdata.get('token')

try:
    client = MilvusClient(
        uri=CLUSTER_ENDPOINT,
        token=API_TOKEN
    )
    print("Success! Connected to Zilliz Cloud.")
except Exception as e:
    print(f"Failed to connect: {e}")


Success! Connected to Zilliz Cloud.


## 8.3 Add the chatbot set up

Now let's add the same backend logic like chroma to set up our chatbot.

In [ ]:
!pip install -q -U sentence-transformers

In [ ]:
from pymilvus import MilvusClient
from sentence_transformers import SentenceTransformer


embedding_model = SentenceTransformer('all-MiniLM-L6-v2')
dimension = embedding_model.get_sentence_embedding_dimension()

collection_name = "answers"

if client.has_collection(collection_name):
    client.drop_collection(collection_name)

client.create_collection(
    collection_name=collection_name,
    dimension=dimension
)

documents = [
    "Le Cookie Quantique is a premium galactic bakery specializing in cosmic-themed treats.",
    "Our best-selling item is the Black-Hole Brownie, which contains no nuts.",
    "The Nebula Nutella Sablé is a fan favorite and contains hazelnuts and chocolate.",
    "We are open Monday through Friday from 08:00 AM to 06:00 PM Earth Time.",
    "All of our Macaron de Mars are naturally gluten-free and made with almond flour.",
    "We offer a fully vegan Supernova Soufflé for our plant-based space travelers.",
    "Our flagship store is located on Earth, but we deliver anywhere in the solar system.",
    "Our Croissant of the Cosmos is baked fresh daily at 4:00 AM to ensure maximum flakiness."
]

queries = [
    "Do you have any vegan cookie options?",
    "What time does the bakery open?"
]


doc_vectors = embedding_model.encode(documents).tolist()
query_vectors = embedding_model.encode(queries).tolist()

insert_data = [
    {"id": i, "vector": doc_vectors[i], "text": documents[i]}
    for i in range(len(documents))
]

client.insert(collection_name=collection_name, data=insert_data)
print(f"Successfully added {len(insert_data)} documents to Milvus!")

print("\n--- Query 1: Vegan Options ---")
res_1 = client.search(
    collection_name=collection_name,
    data=[query_vectors[0]],
    limit=2,
    output_fields=["text"]
)

for hit in res_1[0]:
    print(f"Distance: {hit['distance']:.4f} | Result: {hit['entity']['text']}")

print("\n--- Query 2: Bakery Hours ---")
res_2 = client.search(
    collection_name=collection_name,
    data=[query_vectors[1]],
    limit=2,
    output_fields=["text"]
)

for hit in res_2[0]:
    print(f"Distance: {hit['distance']:.4f} | Result: {hit['entity']['text']}")

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Successfully added 8 documents to Milvus!

--- Query 1: Vegan Options ---
Distance: 0.4360 | Result: We offer a fully vegan Supernova Soufflé for our plant-based space travelers.
Distance: 0.4266 | Result: Le Cookie Quantique is a premium galactic bakery specializing in cosmic-themed treats.

--- Query 2: Bakery Hours ---
Distance: 0.5287 | Result: We are open Monday through Friday from 08:00 AM to 06:00 PM Earth Time.
Distance: 0.3897 | Result: Our Croissant of the Cosmos is baked fresh daily at 4:00 AM to ensure maximum flakiness.


##8.4 So, what happened here?

We used the `sentence-tranformers` model to translate human language into mathematical vectors, We use the sentence-transformers model to translate human language into mathematical vectors, allowing the database to understand the actual meaning of the text rather than just matching keywords. In the code, we first connect to the Milvus database and create a collection to store our data. Next, the model converts our bakery facts and search queries into these number vectors, and we save them into the database alongside the original text. Finally, we run a semantic search where Milvus calculates the mathematical distance between the questions and the stored facts to instantly retrieve the most contextually relevant answers.

## 8.5 Differences between Chroma and Milvus:
 Having implemented the same chatbot in both Chroma and Milvus, below are the key differences I observed -


*  Setup Friction and Developer Experience: Chroma was incredibly easy to set up locally; it practically runs out of the box with a simple pip install and a few basic commands. Milvus, however, has a significantly steeper learning curve and setup process. While milvus-lite exists for local development, it is prone to severe environment conflicts (especially on Windows or within Colab). Ultimately, getting Milvus running reliably required bypassing local execution entirely, provisioning a cluster on Zilliz Cloud, and configuring remote endpoint URIs and API keys just to execute basic database operations.


*  In Chroma, I simply passed raw text to .add(), and the database automatically handled the embedding algorithm, dimensionality, and indexing behind the scenes. In Milvus, the process is much more explicit. I had to manually initialize an embedding model, generate the vector arrays, explicitly define the vector dimensions when creating the collection, and structure the data payload before inserting it.

*   Chroma is fantastic for rapid prototyping, AI hackathons, and small-to-medium applications because it is so lightweight. Milvus, on the other hand, is a cloud-native, highly distributed database built for massive enterprise scale. While milvus allowed me to run it locally for this homework, standard Milvus separates storage from computation across distributed nodes, allowing it to handle billions of vectors and massive query concurrency that would overwhelm a local Chroma instance.

*   Milvus enforces a much stricter schema. When creating the collection, the dimensionality of the vectors must be declared upfront, ensuring data integrity at scale. Chroma's schema-less feel is more flexible but can lead to mismatched vector errors if developers aren't careful when swapping embedding models.